File: java/src/com/algos/topics/li05_binarysearch/revised/L1283_FindtheSmallestDivisorGivenaThreshold.java
package com.algos.topics.li05_binarysearch.revised;

public class L1283_FindtheSmallestDivisorGivenaThreshold {
    public static void main(String[] args) {
        L1283_FindtheSmallestDivisorGivenaThreshold obj = new L1283_FindtheSmallestDivisorGivenaThreshold();
        int[] nums = {1,2,5,9};
        int threshold = 6;
        System.out.println(obj.smallestDivisor(nums, threshold));
    }

    // Problem: https://leetcode.com/problems/find-the-smallest-divisor-given-a-threshold/
    // Idea: Simple BS on range 1 to max element in nums
    // Time: O(nlog(max))
    public int smallestDivisor(int[] nums, int threshold) {
        int max = Integer.MIN_VALUE;
        for(int num: nums)
            max = Math.max(max, num);

        int low = 1, high = max, ans = -1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(sumDivisors(nums, mid) <= threshold){
                ans = mid;
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return ans;
    }

    private int sumDivisors(int[] nums, int mid) {
        int sum = 0;
        for(int num: nums){
            sum += Math.ceil((num*1.0)/mid);
        }
        return sum;
    }
}

File: dsa/patterns/05_binarysearch/python/l1283_findthe_smallest_divisor_givena_threshold.py
"""
LeetCode 1283 - Find the Smallest Divisor Given a Threshold

Idea:
Binary search on divisor.
- Minimum divisor = 1
- Maximum divisor = max(nums)

For a guessed divisor d, compute:
    sum(ceil(num / d)) for all nums
If sum <= threshold, try smaller divisor.
Else, increase divisor.

Time Complexity: O(n log(max(nums)))
Space Complexity: O(1)
"""

import math


def divisor_sum(nums, divisor):
    total = 0
    for num in nums:
        total += math.ceil(num / divisor)
    return total


def smallest_divisor(nums, threshold):
    low, high = 1, max(nums)
    ans = high

    while low <= high:
        mid = (low + high) // 2

        if divisor_sum(nums, mid) <= threshold:
            ans = mid
            high = mid - 1   # try smaller divisor
        else:
            low = mid + 1    # need larger divisor

    return ans


# Driver code
if __name__ == "__main__":
    nums = [1, 2, 5, 9]
    threshold = 6
    print(smallest_divisor(nums, threshold))  # 5
